[rush] rush update can not handle dependenciesMeta in pnpm

<!-- Have a question?  Before creating an issue, ask in the chat room: https://rushstack.zulipchat.com/ -->

<!-- Try invoking Rush with the "--debug" parameter, for example "rush --debug install".  This prints a full call stack when errors occur. -->

<!--------------------------------------------------------------------------
👉 STEP 1: Write a concise but specific issue title in the box above.
     Example: "[rush] Using --example switch causes TypeError"
--------------------------------------------------------------------------->

## Summary
The PNPM has a setting called dependenciesMeta, see more details on [PNPM](https://pnpm.io/package_json#dependenciesmeta)
It is used to set up the injected settings for packages. 

Right now, when users set up this in `package.json`, `rush update` does not update the `pnpm-lock.yaml`

<!--------------------------------------------------------------------------
👉 STEP 2: In a few sentences, please explain:

     What were you trying to accomplish?
     What action did you perform that ran into trouble?
     What went wrong?
--------------------------------------------------------------------------->

## Repro steps

This is a bug I found when I working on [pnpm-sync-prototype](https://github.com/g-chao/pnpm-sync-prototype)
So, we can use this repo to reproduce the bug.

1. Change the injected settings for any workspace dependencies. In this example, let's modify `package.json` for `apps/app3` , add `"dependenciesMeta": {
    "lib1": {
      "injected": true
    }
  }`
3. Run `rush update`
4. Check `pnpm-lock.yaml` under `common/config/rush`, you will see there is **no change** for `pnpm-lock.yaml` 
5. However, if you check `common/temp` folder, you will see `pnpm-lock.yaml` and `pnpm-lock-preinstall.yaml` **are different**
6. Now, delete the `common/temp` folder, and run `rush install`, you will the see PNPM complains `ERR_PNPM_OUTDATED_LOCKFILE  Cannot install with "frozen-lockfile" because pnpm-lock.yaml is not up to date with ../../apps/app3/package.json`


The expected behavior is that after we make some changes to `dependenciesMeta` in `package.json`, `rush update` should update the `pnpm-lock.yaml` under `common/config/rush`. So that, the `rush install` will not fail later. 

<!--------------------------------------------------------------------------
👉 STEP 3: If your issue is a feature request and not a bug, delete this
     "Repro steps" section and skip to STEP 6.

👉 STEP 4: In many cases we can investigate bugs much faster if you include:
     The URL for a simplified Git branch that reproduces the problem.
     Step by step instructions for how to build the branch and see the error.

👉 STEP 5: It's also helpful to include an "expected" and "actual" result.
     But if that's not relevant, feel free to delete those fields.
--------------------------------------------------------------------------->

## Standard questions

Please answer these questions to help us investigate your issue more quickly:

| Question | Answer |
| -------- | -------- |
| `@microsoft/rush` globally installed version? | 5.109.1 |
| `rushVersion` from rush.json? | 5.109.1 |
| `useWorkspaces` from rush.json? | pnpm workspace |
| Operating system? | Mac |
| Would you consider contributing a PR? | Yes |
| Node.js version (`node -v`)? | v16.20.2 |

--- a/common/changes/@microsoft/rush/chao-support-injected-settings_2023-12-13-18-48.json
+++ b/common/changes/@microsoft/rush/chao-support-injected-settings_2023-12-13-18-48.json
@@ -2,7 +2,7 @@
   "changes": [
     {
       "packageName": "@microsoft/rush",
-      "comment": "Fix an issue where rush update can not handle dependenciesMeta changes in pnpm",
+      "comment": "Fix an issue where Rush does not detect changes to the `dependenciesMeta` field in project's `package.json` files, so may incorrectly skip updating/installation.",
       "type": "none"
     }
   ],

--- a/common/changes/@microsoft/rush/chao-support-injected-settings_2023-12-13-18-48.json
+++ b/common/changes/@microsoft/rush/chao-support-injected-settings_2023-12-13-18-48.json
@@ -1,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "@microsoft/rush",
+      "comment": "Fix an issue where rush update can not handle dependenciesMeta changes in pnpm",
+      "type": "none"
+    }
+  ],
+  "packageName": "@microsoft/rush"
+}

--- a/common/changes/@rushstack/node-core-library/chao-support-injected-settings_2023-12-13-18-48.json
+++ b/common/changes/@rushstack/node-core-library/chao-support-injected-settings_2023-12-13-18-48.json
@@ -1,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "@rushstack/node-core-library",
+      "comment": "Add dependenciesMeta interface in packageJson",
+      "type": "patch"
+    }
+  ],
+  "packageName": "@rushstack/node-core-library"
+}

--- a/common/changes/@rushstack/node-core-library/chao-support-injected-settings_2023-12-13-18-48.json
+++ b/common/changes/@rushstack/node-core-library/chao-support-injected-settings_2023-12-13-18-48.json
@@ -2,8 +2,8 @@
   "changes": [
     {
       "packageName": "@rushstack/node-core-library",
-      "comment": "Add dependenciesMeta interface in packageJson",
-      "type": "patch"
+      "comment": "Add the `dependenciesMeta` property to the `INodePackageJson` interface.",
+      "type": "minor"
     }
   ],
   "packageName": "@rushstack/node-core-library"

--- a/common/reviews/api/node-core-library.api.md
+++ b/common/reviews/api/node-core-library.api.md
@@ -553,7 +553,7 @@ export class Import {
 export interface INodePackageJson {
     bin?: string;
     dependencies?: IPackageJsonDependencyTable;
-    dependenciesMeta?: IPeerDependenciesMetaTable;
+    dependenciesMeta?: IDependenciesMetaTable;
     description?: string;
     devDependencies?: IPackageJsonDependencyTable;
     homepage?: string;

--- a/common/reviews/api/node-core-library.api.md
+++ b/common/reviews/api/node-core-library.api.md
@@ -340,6 +340,14 @@ export interface IConsoleTerminalProviderOptions {
     verboseEnabled: boolean;
 }
 
+// @public
+export interface IDependenciesMetaTable {
+    // (undocumented)
+    [dependencyName: string]: {
+        injected?: boolean;
+    };
+}
+
 // @beta
 export interface IDynamicPrefixProxyTerminalProviderOptions extends IPrefixProxyTerminalProviderOptionsBase {
     getPrefix: () => string;
@@ -545,6 +553,7 @@ export class Import {
 export interface INodePackageJson {
     bin?: string;
     dependencies?: IPackageJsonDependencyTable;
+    dependenciesMeta?: IPeerDependenciesMetaTable;
     description?: string;
     devDependencies?: IPackageJsonDependencyTable;
     homepage?: string;

--- a/common/reviews/api/rush-lib.api.md
+++ b/common/reviews/api/rush-lib.api.md
@@ -902,12 +902,22 @@ export class PackageJsonDependency {
 }
 
 // @public (undocumented)
+export class PackageJsonDependencyMeta {
+    constructor(name: string, injected: boolean, onChange: () => void);
+    // (undocumented)
+    get injected(): boolean;
+    // (undocumented)
+    readonly name: string;
+}
+
+// @public (undocumented)
 export class PackageJsonEditor {
     // @internal
     protected constructor(filepath: string, data: IPackageJson);
     // (undocumented)
     addOrUpdateDependency(packageName: string, newVersion: string, dependencyType: DependencyType): void;
     get dependencyList(): ReadonlyArray<PackageJsonDependency>;
+    get dependencyMetaList(): ReadonlyArray<PackageJsonDependencyMeta>;
     get devDependencyList(): ReadonlyArray<PackageJsonDependency>;
     // (undocumented)
     readonly filePath: string;

--- a/libraries/node-core-library/src/IPackageJson.ts
+++ b/libraries/node-core-library/src/IPackageJson.ts
@@ -61,6 +61,17 @@ export interface IPeerDependenciesMetaTable {
 }
 
 /**
+ * This interface is part of the {@link IPackageJson} file format. It is used for the
+ * "dependenciesMeta" field.
+ * @public
+ */
+export interface IDependenciesMetaTable {
+  [dependencyName: string]: {
+    injected?: boolean;
+  };
+}
+
+/**
  * An interface for accessing common fields from a package.json file whose version field may be missing.
  *
  * @remarks
@@ -165,6 +176,12 @@ export interface INodePackageJson {
    * but which will not be automatically installed by this package.
    */
   peerDependencies?: IPackageJsonDependencyTable;
+
+  /**
+   * An array of metadata for dependencies declared inside dependencies, optionalDependencies, and devDependencies.
+   * https://pnpm.io/package_json#dependenciesmeta
+   */
+  dependenciesMeta?: IPeerDependenciesMetaTable;
 
   /**
    * An array of metadata about peer dependencies.

--- a/libraries/node-core-library/src/IPackageJson.ts
+++ b/libraries/node-core-library/src/IPackageJson.ts
@@ -181,7 +181,7 @@ export interface INodePackageJson {
    * An array of metadata for dependencies declared inside dependencies, optionalDependencies, and devDependencies.
    * https://pnpm.io/package_json#dependenciesmeta
    */
-  dependenciesMeta?: IPeerDependenciesMetaTable;
+  dependenciesMeta?: IDependenciesMetaTable;
 
   /**
    * An array of metadata about peer dependencies.

--- a/libraries/node-core-library/src/index.ts
+++ b/libraries/node-core-library/src/index.ts
@@ -29,7 +29,8 @@ export {
   IPackageJsonDependencyTable,
   IPackageJsonScriptTable,
   IPackageJsonRepository,
-  IPeerDependenciesMetaTable
+  IPeerDependenciesMetaTable,
+  IDependenciesMetaTable
 } from './IPackageJson';
 export {
   Import,

--- a/libraries/rush-lib/src/api/PackageJsonEditor.ts
+++ b/libraries/rush-lib/src/api/PackageJsonEditor.ts
@@ -49,6 +49,26 @@ export class PackageJsonDependency {
 /**
  * @public
  */
+export class PackageJsonDependencyMeta {
+  private _injected: boolean;
+  private _onChange: () => void;
+
+  public readonly name: string;
+
+  public constructor(name: string, injected: boolean, onChange: () => void) {
+    this.name = name;
+    this._injected = injected;
+    this._onChange = onChange;
+  }
+
+  public get injected(): boolean {
+    return this._injected;
+  }
+}
+
+/**
+ * @public
+ */
 export class PackageJsonEditor {
   private readonly _dependencies: Map<string, PackageJsonDependency>;
   // NOTE: The "devDependencies" section is tracked separately because sometimes people
@@ -56,6 +76,8 @@ export class PackageJsonEditor {
   // SemVer range in one of the other fields for consumers.  Thus "dependencies", "optionalDependencies",
   // and "peerDependencies" are mutually exclusive, but "devDependencies" is not.
   private readonly _devDependencies: Map<string, PackageJsonDependency>;
+
+  private readonly _dependenciesMeta: Map<string, PackageJsonDependencyMeta>;
 
   // NOTE: The "resolutions" field is a yarn specific feature that controls package
   // resolution override within yarn.
@@ -76,6 +98,7 @@ export class PackageJsonEditor {
     this._dependencies = new Map<string, PackageJsonDependency>();
     this._devDependencies = new Map<string, PackageJsonDependency>();
     this._resolutions = new Map<string, PackageJsonDependency>();
+    this._dependenciesMeta = new Map<string, PackageJsonDependencyMeta>();
 
     const dependencies: { [key: string]: string } = data.dependencies || {};
     const optionalDependencies: { [key: string]: string } = data.optionalDependencies || {};
@@ -83,6 +106,8 @@ export class PackageJsonEditor {
 
     const devDependencies: { [key: string]: string } = data.devDependencies || {};
     const resolutions: { [key: string]: string } = data.resolutions || {};
+
+    const dependenciesMeta: { [key: string]: { [key: string]: boolean } } = data.dependenciesMeta || {};
 
     const _onChange: () => void = this._onChange.bind(this);
 
@@ -155,6 +180,13 @@ export class PackageJsonEditor {
         );
       });
 
+      Object.keys(dependenciesMeta || {}).forEach((packageName: string) => {
+        this._dependenciesMeta.set(
+          packageName,
+          new PackageJsonDependencyMeta(packageName, dependenciesMeta[packageName].injected, _onChange)
+        );
+      });
+
       // (Do not sort this._resolutions because order may be significant; the RFC is unclear about that.)
       Sort.sortMapKeys(this._dependencies);
       Sort.sortMapKeys(this._devDependencies);
@@ -191,6 +223,13 @@ export class PackageJsonEditor {
    */
   public get devDependencyList(): ReadonlyArray<PackageJsonDependency> {
     return [...this._devDependencies.values()];
+  }
+
+  /**
+   * The list of dependenciesMeta in package.json.
+   */
+  public get dependencyMetaList(): ReadonlyArray<PackageJsonDependencyMeta> {
+    return [...this._dependenciesMeta.values()];
   }
 
   /**

--- a/libraries/rush-lib/src/index.ts
+++ b/libraries/rush-lib/src/index.ts
@@ -71,7 +71,12 @@ export { ApprovedPackagesItem, ApprovedPackagesConfiguration } from './api/Appro
 
 export { CommonVersionsConfiguration } from './api/CommonVersionsConfiguration';
 
-export { PackageJsonEditor, PackageJsonDependency, DependencyType } from './api/PackageJsonEditor';
+export {
+  PackageJsonEditor,
+  PackageJsonDependency,
+  DependencyType,
+  PackageJsonDependencyMeta
+} from './api/PackageJsonEditor';
 
 export { RepoStateFile } from './logic/RepoStateFile';
 

--- a/libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts
+++ b/libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts
@@ -256,13 +256,13 @@ export class WorkspaceInstallManager extends BaseInstallManager {
       const commonTempFolder: string = this.rushConfiguration.commonTempFolder;
       const rushJsonFolder: string = this.rushConfiguration.rushJsonFolder;
 
-      // get the relative package path from common temp folder to repo root folder
+      // get the relative path from common temp folder to repo root folder
       const relativeFromTempFolderToRootFolder: string = path.relative(commonTempFolder, rushJsonFolder);
       for (const rushProject of this.rushConfiguration.projects) {
         const packageJson: PackageJsonEditor = rushProject.packageJsonEditor;
         const projectRelativeFolder: string = rushProject.projectRelativeFolder;
 
-        // get the relative package path from common temp folder to package folder, to align with the value in pnpm-lock.yaml
+        // get the relative path from common temp folder to package folder, to align with the value in pnpm-lock.yaml
         const relativePathFromTempFolderToPackageFolder: string =
           relativeFromTempFolderToRootFolder + '/' + projectRelativeFolder;
         const dependencyMetaList: ReadonlyArray<PackageJsonDependencyMeta> = packageJson.dependencyMetaList;

--- a/libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts
+++ b/libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts
@@ -4,13 +4,23 @@
 import colors from 'colors/safe';
 import * as path from 'path';
 import * as semver from 'semver';
-import { FileSystem, FileConstants, AlreadyReportedError, Async } from '@rushstack/node-core-library';
+import {
+  FileSystem,
+  FileConstants,
+  AlreadyReportedError,
+  Async,
+  type IDependenciesMetaTable
+} from '@rushstack/node-core-library';
 
 import { BaseInstallManager } from '../base/BaseInstallManager';
 import type { IInstallManagerOptions } from '../base/BaseInstallManagerTypes';
 import type { BaseShrinkwrapFile } from '../../logic/base/BaseShrinkwrapFile';
 import { DependencySpecifier, DependencySpecifierType } from '../DependencySpecifier';
-import { type PackageJsonEditor, DependencyType } from '../../api/PackageJsonEditor';
+import {
+  type PackageJsonEditor,
+  DependencyType,
+  type PackageJsonDependencyMeta
+} from '../../api/PackageJsonEditor';
 import { PnpmWorkspaceFile } from '../pnpm/PnpmWorkspaceFile';
 import type { RushConfigurationProject } from '../../api/RushConfigurationProject';
 import { RushConstants } from '../../logic/RushConstants';
@@ -23,6 +33,8 @@ import { EnvironmentConfiguration } from '../../api/EnvironmentConfiguration';
 import { ShrinkwrapFileFactory } from '../ShrinkwrapFileFactory';
 import { BaseProjectShrinkwrapFile } from '../base/BaseProjectShrinkwrapFile';
 import { type CustomTipId, type ICustomTipInfo, PNPM_CUSTOM_TIPS } from '../../api/CustomTipsConfiguration';
+import type { PnpmShrinkwrapFile } from '../pnpm/PnpmShrinkwrapFile';
+import { objectsAreDeepEqual } from '../../utilities/objectUtilities';
 
 /**
  * This class implements common logic between "rush install" and "rush update".
@@ -56,7 +68,7 @@ export class WorkspaceInstallManager extends BaseInstallManager {
    * @override
    */
   protected async prepareCommonTempAsync(
-    shrinkwrapFile: BaseShrinkwrapFile | undefined
+    shrinkwrapFile: (PnpmShrinkwrapFile & BaseShrinkwrapFile) | undefined
   ): Promise<{ shrinkwrapIsUpToDate: boolean; shrinkwrapWarnings: string[] }> {
     // Block use of the RUSH_TEMP_FOLDER environment variable
     if (EnvironmentConfiguration.rushTempFolderOverride !== undefined) {
@@ -233,6 +245,54 @@ export class WorkspaceInstallManager extends BaseInstallManager {
         );
         shrinkwrapIsUpToDate = false;
       }
+    }
+
+    // For pnpm pacakge manager, we need to handle dependenciesMeta changes in package.json. See more: https://pnpm.io/package_json#dependenciesmeta
+    // If dependenciesMeta settings is different between package.json and pnpm-lock.yaml, then shrinkwrapIsUpToDate return false.
+    if (this.rushConfiguration.packageManager === 'pnpm') {
+      // First build a object for dependenciesMeta settings in package.json
+      // key is the package path, value is the dependenciesMeta info for that package
+      const packagePathToDependenciesMetaInPackageJson: { [key: string]: IDependenciesMetaTable } = {};
+      const commonTempFolder: string = this.rushConfiguration.commonTempFolder;
+
+      for (const rushProject of this.rushConfiguration.projects) {
+        const packageJson: PackageJsonEditor = rushProject.packageJsonEditor;
+        const packageJsonFilePath: string = packageJson.filePath;
+
+        // get the relativePackagePath for a package, to align with the value in pnpm-lock.yaml
+        const relativePackagePath: string = path.relative(
+          commonTempFolder,
+          packageJsonFilePath.replace('/package.json', '')
+        );
+        const dependencyMetaList: ReadonlyArray<PackageJsonDependencyMeta> = packageJson.dependencyMetaList;
+
+        if (dependencyMetaList.length !== 0) {
+          const dependenciesMeta: IDependenciesMetaTable = {};
+          for (const dependencyMeta of dependencyMetaList) {
+            dependenciesMeta[dependencyMeta.name] = {
+              injected: dependencyMeta.injected
+            };
+          }
+          packagePathToDependenciesMetaInPackageJson[relativePackagePath] = dependenciesMeta;
+        }
+      }
+
+      // First build a object for dependenciesMeta settings in pnpm-lock.yaml
+      // key is the package path, value is the dependenciesMeta info for that package
+      const packagePathToDependenciesMetaInShrinkwrapFile: { [key: string]: IDependenciesMetaTable } = {};
+      if (shrinkwrapFile?.importers !== undefined) {
+        for (const [key, value] of shrinkwrapFile?.importers) {
+          if (value.dependenciesMeta !== undefined) {
+            packagePathToDependenciesMetaInShrinkwrapFile[key] = value.dependenciesMeta;
+          }
+        }
+      }
+
+      // Now, we compare these two objects to see if they are equal or not
+      shrinkwrapIsUpToDate = objectsAreDeepEqual(
+        packagePathToDependenciesMetaInPackageJson,
+        packagePathToDependenciesMetaInShrinkwrapFile
+      );
     }
 
     // Write the common package.json

--- a/libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts
+++ b/libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts
@@ -250,20 +250,21 @@ export class WorkspaceInstallManager extends BaseInstallManager {
     // For pnpm pacakge manager, we need to handle dependenciesMeta changes in package.json. See more: https://pnpm.io/package_json#dependenciesmeta
     // If dependenciesMeta settings is different between package.json and pnpm-lock.yaml, then shrinkwrapIsUpToDate return false.
     if (this.rushConfiguration.packageManager === 'pnpm') {
-      // First build a object for dependenciesMeta settings in package.json
+      // First, build a object for dependenciesMeta settings in package.json
       // key is the package path, value is the dependenciesMeta info for that package
       const packagePathToDependenciesMetaInPackageJson: { [key: string]: IDependenciesMetaTable } = {};
       const commonTempFolder: string = this.rushConfiguration.commonTempFolder;
+      const rushJsonFolder: string = this.rushConfiguration.rushJsonFolder;
 
+      // get the relative package path from common temp folder to repo root folder
+      const relativeFromTempFolderToRootFolder: string = path.relative(commonTempFolder, rushJsonFolder);
       for (const rushProject of this.rushConfiguration.projects) {
         const packageJson: PackageJsonEditor = rushProject.packageJsonEditor;
-        const packageJsonFilePath: string = packageJson.filePath;
+        const projectRelativeFolder: string = rushProject.projectRelativeFolder;
 
-        // get the relativePackagePath for a package, to align with the value in pnpm-lock.yaml
-        const relativePackagePath: string = path.relative(
-          commonTempFolder,
-          packageJsonFilePath.replace('/package.json', '')
-        );
+        // get the relative package path from common temp folder to package folder, to align with the value in pnpm-lock.yaml
+        const relativePathFromTempFolderToPackageFolder: string =
+          relativeFromTempFolderToRootFolder + '/' + projectRelativeFolder;
         const dependencyMetaList: ReadonlyArray<PackageJsonDependencyMeta> = packageJson.dependencyMetaList;
 
         if (dependencyMetaList.length !== 0) {
@@ -273,11 +274,12 @@ export class WorkspaceInstallManager extends BaseInstallManager {
               injected: dependencyMeta.injected
             };
           }
-          packagePathToDependenciesMetaInPackageJson[relativePackagePath] = dependenciesMeta;
+          packagePathToDependenciesMetaInPackageJson[relativePathFromTempFolderToPackageFolder] =
+            dependenciesMeta;
         }
       }
 
-      // First build a object for dependenciesMeta settings in pnpm-lock.yaml
+      // Second, build a object for dependenciesMeta settings in pnpm-lock.yaml
       // key is the package path, value is the dependenciesMeta info for that package
       const packagePathToDependenciesMetaInShrinkwrapFile: { [key: string]: IDependenciesMetaTable } = {};
       if (shrinkwrapFile?.importers !== undefined) {

--- a/libraries/rush-lib/src/logic/pnpm/PnpmShrinkwrapFile.ts
+++ b/libraries/rush-lib/src/logic/pnpm/PnpmShrinkwrapFile.ts
@@ -33,6 +33,9 @@ const yamlModule: typeof import('js-yaml') = Import.lazy('js-yaml', require);
 export interface IPeerDependenciesMetaYaml {
   optional?: boolean;
 }
+export interface IDependenciesMetaYaml {
+  injected?: boolean;
+}
 
 export type IPnpmV7VersionSpecifier = string;
 export interface IPnpmV8VersionSpecifier {
@@ -69,6 +72,8 @@ export interface IPnpmShrinkwrapImporterYaml {
   devDependencies?: Record<string, IPnpmVersionSpecifier>;
   /** The list of resolved version numbers for optional dependencies */
   optionalDependencies?: Record<string, IPnpmVersionSpecifier>;
+  /** The list of metadata for dependencies declared inside dependencies, optionalDependencies, and devDependencies. */
+  dependenciesMeta?: Record<string, IDependenciesMetaYaml>;
   /**
    * The list of specifiers used to resolve dependency versions
    *

